src/mn_setup.c

```diff
@@ -1185,52 +1221,64 @@ static void DrawInstructions()
         }
         else if (flags & S_RESET)
         {
-            s = "Restore defaults";
+            if (pad)
+            {
+                first = M_GetPlatformName(GAMEPAD_A);
+                second = M_GetPlatformName(GAMEPAD_B);
+            }
+            else
+            {
+                first = M_GetNameForKey(KEY_ENTER);
+                second = M_GetNameForKey(KEY_ESCAPE);
+            }
+
+            HelpString(&s, "[ %s ] OK, [ %s ] Cancel", first, second);
```

Comment:
Another way to do it:
```diff
diff --git a/src/mn_setup.c b/src/mn_setup.c
index dc6c7213..ca5c96b7 100644
--- a/src/mn_setup.c
+++ b/src/mn_setup.c
@@ -1131,16 +1131,6 @@ typedef enum
 
 static menu_help_t menu_help;
 
-static void PRINTF_ATTR(2, 3) HelpString(const char **s, const char *fmt, ...)
-{
-    static char buf[64];
-    va_list args;
-    va_start(args, fmt);
-    M_vsnprintf(buf, sizeof(buf), fmt, args);
-    va_end(args);
-    *s = buf;
-}
-
 static void DrawInstructions()
 {
     int index = (menu_input == mouse_mode ? highlight_item : set_item_on);
@@ -1155,7 +1145,7 @@ static void DrawInstructions()
     // There are different instruction messages depending on whether you
     // are changing an item or just sitting on it.
 
-    const char *s = "";
+    char *s = NULL;
     const char *first, *second;
     const boolean pad = ((menu_help == MENU_HELP_AUTO && help_input == pad_mode)
                          || menu_help == MENU_HELP_PAD);
@@ -1171,17 +1161,17 @@ static void DrawInstructions()
             second = M_GetNameForKey(KEY_BACKSPACE);
         }
 
-        HelpString(&s, "[ %s ] Back", second);
+        s = M_StringJoin("[", second, "] Back");
     }
     else if (item->desc)
     {
-        s = item->desc;
+        s = M_StringDuplicate(item->desc);
     }
     else if (setup_select)
     {
         if (flags & S_INPUT)
         {
-            s = "Press key or button to bind/unbind";
+            s = M_StringDuplicate("Press key or button to bind/unbind");
         }
         else if (flags & S_ONOFF)
         {
@@ -1196,7 +1186,7 @@ static void DrawInstructions()
                 second = M_GetNameForKey(KEY_ESCAPE);
             }
 
-            HelpString(&s, "[ %s ] Toggle, [ %s ] Cancel", first, second);
+            s = M_StringJoin("[", first, "] Toggle, [", second, "] Cancel");
         }
         else if (flags & (S_CHOICE | S_CRITEM | S_THERMO))
         {
@@ -1209,15 +1199,15 @@ static void DrawInstructions()
                 second = M_GetNameForKey(KEY_ESCAPE);
             }
 
-            HelpString(&s, "[ Left/Right ] Choose, [ %s ] Cancel", second);
+            s = M_StringJoin("[ Left/Right ] Choose, [", second, "] Cancel");
         }
         else if (flags & S_NUM)
         {
-            s = "Enter value";
+            s = M_StringDuplicate("Enter value");
         }
         else if (flags & S_WEAP)
         {
-            s = "Enter weapon number";
+            s = M_StringDuplicate("Enter weapon number");
         }
         else if (flags & S_RESET)
         {
@@ -1232,7 +1222,7 @@ static void DrawInstructions()
                 second = M_GetNameForKey(KEY_ESCAPE);
             }
 
-            HelpString(&s, "[ %s ] OK, [ %s ] Cancel", first, second);
+            s = M_StringJoin("[", first, "] OK, [", second, "] Cancel");
         }
     }
     else
@@ -1250,7 +1240,7 @@ static void DrawInstructions()
                 second = M_GetNameForKey(KEY_DEL);
             }
 
-            HelpString(&s, "[ %s ] Change, [ %s ] Clear", first, second);
+            s = M_StringJoin("[", first, "] Change, [", second,"] Clear");
         }
         else if (flags & S_RESET)
         {
@@ -1263,7 +1253,7 @@ static void DrawInstructions()
                 first = M_GetNameForKey(KEY_ENTER);
             }
 
-            HelpString(&s, "[ %s ] Restore defaults", first);
+            s = M_StringJoin("[", first, "] Restore defaults");
         }
         else
         {
@@ -1278,12 +1268,16 @@ static void DrawInstructions()
                 second = M_GetNameForKey(KEY_BACKSPACE);
             }
 
-            HelpString(&s, "[ %s ] Change, [ %s ] Back", first, second);
+            s = M_StringJoin("[", first, "] Change, [", second, "] Back");
         }
     }
 
-    MN_DrawStringCR((SCREENWIDTH - MN_GetPixelWidth(s)) / 2, M_Y_WARN,
-                    cr_shaded, NULL, s);
+    if (s)
+    {
+        MN_DrawStringCR((SCREENWIDTH - MN_GetPixelWidth(s)) / 2, M_Y_WARN,
+                        cr_shaded, NULL, s);
+        free(s);
+    }
 }
 
 static void SetupMenu(void)
```

Reply:
I like it! Then we don't have to deal with the va_list stuff.

Reply:
Another way would be to create a static buffer (our space is limited anyway) and `M_snprintf` into it.
```c
char s[80];
...
M_snprintf(s, sizeof(s), "[%s] Toggle, [%s] Cancel", first, second);
```
> Then we don't have to deal with the va_list stuff.

Yes, I think we should avoid it.

Reply:
Fixed by https://github.com/fabiangreffrath/woof/pull/1890/commits/cb54c74ac5c0fb4d440624015d41a64fccb4c49d